File: engine/src/flutter/lib/gpu/lib/src/texture.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: public_member_api_docs

part of flutter_gpu;

base class Texture extends NativeFieldWrapperClass1 {
  bool _valid = false;
  get isValid {
    return _valid;
  }

  /// Returns the maximum number of mip levels Flutter GPU will allocate
  /// for a texture of the given [width] and [height].
  ///
  /// This currently uses `floor(log2(min(w, h)))` and clamps at 1. Note
  /// that this is one less than the canonical `floor(log2(max(w, h))) + 1`
  /// used by Vulkan and OpenGL, and uses the smaller dimension instead of
  /// the larger; Flutter GPU may eventually relax this (tracked in
  /// https://github.com/flutter/flutter/issues/186176). Users should treat
  /// this as the upper bound for the `mipLevelCount` argument to
  /// `GpuContext.createTexture` rather than as a fixed mathematical
  /// formula.
  static int fullMipCount(int width, int height) {
    if (width < 1 || height < 1) {
      return 1;
    }
    final int smallest = width < height ? width : height;
    final int count = smallest.bitLength - 1;
    return count > 0 ? count : 1;
  }

  /// Creates a new Texture.
  Texture._initialize(
    GpuContext gpuContext,
    this.storageMode,
    this.format,
    this.width,
    this.height,
    this.sampleCount,
    TextureCoordinateSystem coordinateSystem,
    this.textureType,
    this.enableRenderTargetUsage,
    this.enableShaderReadUsage,
    this.enableShaderWriteUsage,
    this.mipLevelCount,
  ) : _gpuContext = gpuContext,
      _coordinateSystem = coordinateSystem {
    if (sampleCount != 1 && sampleCount != 4) {
      throw Exception("Only a sample count of 1 or 4 is currently supported");
    }
    _valid = _initialize(
      gpuContext,
      storageMode.index,
      format.index,
      width,
      height,
      sampleCount,
      coordinateSystem.index,
      textureType.index,
      enableRenderTargetUsage,
      enableShaderReadUsage,
      enableShaderWriteUsage,
      mipLevelCount,
    );
  }

  GpuContext _gpuContext;

  final StorageMode storageMode;
  final PixelFormat format;
  final int width;
  final int height;
  final int sampleCount;
  final TextureType textureType;

  /// Enable using this texture as a render pass attachment.
  final bool enableRenderTargetUsage;

  /// Enable reading or sampling from this texture in a shader.
  final bool enableShaderReadUsage;

  /// Enable writing to the texture in a shader.
  ///
  /// Note that this is distinct from [enableRenderTargetUsage].
  final bool enableShaderWriteUsage;

  /// The number of mip levels allocated for this texture.
  ///
  /// A value of 1 means no mip chain (only the base level). Larger values
  /// allocate additional levels with halved dimensions per step. Use
  /// [Texture.fullMipCount] to compute the maximum for a given size.
  final int mipLevelCount;

  /// The number of slices in this texture. Determined by [textureType]:
  /// 1 for 2D and external textures, 6 for cubemap textures.
  int get sliceCount => textureType == TextureType.textureCube ? 6 : 1;

  TextureCoordinateSystem _coordinateSystem;
  TextureCoordinateSystem get coordinateSystem {
    return _coordinateSystem;
  }

  set coordinateSystem(TextureCoordinateSystem value) {
    value;
    _setCoordinateSystem(value.index);
  }

  int get bytesPerTexel {
    return _bytesPerTexel();
  }

  /// Returns the size in bytes of the [mipLevel] mip level (one slice). Mip
  /// dimensions are clamped at 1, matching standard mip chain semantics.
  int getMipLevelSizeInBytes(int mipLevel) {
    final int mipWidth = width >> mipLevel;
    final int mipHeight = height >> mipLevel;
    final int w = mipWidth > 0 ? mipWidth : 1;
    final int h = mipHeight > 0 ? mipHeight : 1;
    return bytesPerTexel * w * h;
  }

  /// Returns the size in bytes of the base mip level (one slice). Equivalent
  /// to `getMipLevelSizeInBytes(0)`.
  int getBaseMipLevelSizeInBytes() {
    return getMipLevelSizeInBytes(0);
  }

  /// Overwrites a mip level (and slice for cubemap textures) of this
  /// [Texture] with the contents of [sourceBytes].
  ///
  /// [mipLevel] selects which mip level to write to. Defaults to 0 (base
  /// level). Must be in the range `[0, mipLevelCount)`.
  ///
  /// [slice] selects which slice to write to for cubemap textures, where
  /// each face is a separate slice in the order
  /// `+X, -X, +Y, -Y, +Z, -Z`. Must be 0 for non-cubemap textures.
  ///
  /// The length of [sourceBytes] must exactly match the size of the
  /// requested mip level, which is `mipWidth * mipHeight * bytesPerTexel`
  /// (where `mipWidth` and `mipHeight` are the base dimensions right-shifted
  /// by [mipLevel], floored at 1).
  ///
  /// Throws an exception if the write fails due to an internal error or if
  /// any of the parameters are out of range.
  void overwrite(ByteData sourceBytes, {int mipLevel = 0, int slice = 0}) {
    if (mipLevel < 0 || mipLevel >= mipLevelCount) {
      throw Exception(
        'mipLevel ($mipLevel) must be in the range [0, $mipLevelCount) for this texture',
      );
    }
    final int slices = sliceCount;
    if (slice < 0 || slice >= slices) {
      throw Exception(
        'slice ($slice) must be in the range [0, $slices) for textures of type $textureType',
      );
    }
    final int expectedSize = getMipLevelSizeInBytes(mipLevel);
    if (sourceBytes.lengthInBytes != expectedSize) {
      throw Exception(
        'The length of sourceBytes (bytes: ${sourceBytes.lengthInBytes}) must exactly match the size of mip level $mipLevel (bytes: $expectedSize)',
      );
    }
    bool success = _overwrite(_gpuContext, sourceBytes, mipLevel, slice);
    if (!success) {
      throw Exception("Texture overwrite failed");
    }
  }

  ui.Image asImage() {
    if (!enableShaderReadUsage) {
      throw Exception(
        'Only shader readable Flutter GPU textures can be used as UI Images',
      );
    }
    return _asImage();
  }

  /// Wrap with native counterpart.
  @Native<
    Bool Function(
      Handle,
      Pointer<Void>,
      Int,
      Int,
      Int,
      Int,
      Int,
      Int,
      Int,
      Bool,
      Bool,
      Bool,
      Int,
    )
  >(symbol: 'InternalFlutterGpu_Texture_Initialize')
  external bool _initialize(
    GpuContext gpuContext,
    int storageMode,
    int format,
    int width,
    int height,
    int sampleCount,
    int coordinateSystem,
    int textureType,
    bool enableRenderTargetUsage,
    bool enableShaderReadUsage,
    bool enableShaderWriteUsage,
    int mipLevelCount,
  );

  @Native<Void Function(Handle, Int)>(
    symbol: 'InternalFlutterGpu_Texture_SetCoordinateSystem',
  )
  external void _setCoordinateSystem(int coordinateSystem);

  @Native<Int Function(Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_Texture_BytesPerTexel',
  )
  external int _bytesPerTexel();

  @Native<Bool Function(Pointer<Void>, Pointer<Void>, Handle, Int, Int)>(
    symbol: 'InternalFlutterGpu_Texture_Overwrite',
  )
  external bool _overwrite(
    GpuContext gpuContext,
    ByteData bytes,
    int mipLevel,
    int slice,
  );

  @Native<Handle Function(Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_Texture_AsImage',
  )
  external ui.Image _asImage();
}

File: engine/src/flutter/lib/gpu/texture.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/gpu/texture.h"

#include <cstring>

#include "flutter/lib/gpu/formats.h"
#include "flutter/lib/ui/painting/image.h"
#include "flutter/lib/ui/ui_dart_state.h"
#include "fml/make_copyable.h"
#include "fml/mapping.h"
#include "impeller/core/allocator.h"
#include "impeller/core/buffer_view.h"
#include "impeller/core/device_buffer.h"
#include "impeller/core/formats.h"
#include "impeller/core/texture.h"
#include "impeller/geometry/rect.h"
#include "impeller/renderer/blit_pass.h"
#include "impeller/renderer/command_buffer.h"
#include "impeller/renderer/command_queue.h"
#include "impeller/renderer/context.h"

#if IMPELLER_SUPPORTS_RENDERING
#include "flutter/display_list/image/dl_image.h"      // nogncheck
#include "impeller/display_list/dl_image_impeller.h"  // nogncheck
#endif
#include "third_party/tonic/converter/dart_converter.h"
#include "third_party/tonic/dart_wrappable.h"
#include "third_party/tonic/typed_data/dart_byte_data.h"

namespace flutter {
namespace gpu {

IMPLEMENT_WRAPPERTYPEINFO(flutter_gpu, Texture);

Texture::Texture(std::shared_ptr<impeller::Texture> texture)
    : texture_(std::move(texture)) {}

Texture::~Texture() = default;

std::shared_ptr<impeller::Texture> Texture::GetTexture() {
  return texture_;
}

// Returns the size in pixels of the given dimension at `mip_level`, clamped
// at 1, matching standard mip-chain semantics. The Dart-side helper
// `Texture.getMipLevelSizeInBytes` uses the same `max(1, dim >> level)`
// formula in pixel-count form; if either is updated, both must be kept in
// sync.
static int32_t MipDimensionAtLevel(int32_t base_dimension, uint32_t mip_level) {
  const int32_t shifted = base_dimension >> mip_level;
  return shifted > 0 ? shifted : 1;
}

// Records a blit-pass that copies `source_bytes` into the given mip level and
// slice of `texture` on `context`, then submits the command buffer. Returns
// true if the encode and submit both succeed. The actual GPU upload may
// complete asynchronously after this call returns.
static bool EncodeAndSubmitOverwrite(
    impeller::Context& context,
    const std::shared_ptr<impeller::Texture>& texture,
    const std::shared_ptr<impeller::DeviceBuffer>& staging_buffer,
    size_t source_length,
    impeller::IRect destination_region,
    uint32_t mip_level,
    uint32_t slice) {
  auto command_buffer = context.CreateCommandBuffer();
  if (!command_buffer) {
    FML_LOG(ERROR) << "Failed to create command buffer for texture overwrite.";
    return false;
  }
  auto blit_pass = command_buffer->CreateBlitPass();
  if (!blit_pass) {
    FML_LOG(ERROR) << "Failed to create blit pass for texture overwrite.";
    return false;
  }
  impeller::BufferView buffer_view(staging_buffer,
                                   impeller::Range(0, source_length));
  if (!blit_pass->AddCopy(std::move(buffer_view), texture, destination_region,
                          /*label=*/"Texture.overwrite", mip_level, slice)) {
    return false;
  }
  if (!blit_pass->EncodeCommands()) {
    return false;
  }
  return context.GetCommandQueue()->Submit({std::move(command_buffer)}).ok();
}

bool Texture::Overwrite(Context& gpu_context,
                        const tonic::DartByteData& source_bytes,
                        uint32_t mip_level,
                        uint32_t slice) {
  const uint8_t* data = static_cast<const uint8_t*>(source_bytes.data());
  const size_t length = source_bytes.length_in_bytes();

  auto& impeller_context = gpu_context.GetContext();
  auto staging_buffer =
      impeller_context.GetResourceAllocator()->CreateBufferWithCopy(data,
                                                                    length);
  if (!staging_buffer) {
    FML_LOG(ERROR) << "Failed to allocate staging buffer for texture "
                      "overwrite.";
    return false;
  }

  // Compute the destination region for the requested mip level. The
  // BlitPass::AddCopy validation requires the region to fit within the base
  // texture size, and the actual GPU copy uses this rectangle as the
  // destination on the chosen mip level. The same `max(1, dim >> level)`
  // formula is used by `Texture.getMipLevelSizeInBytes` on the Dart side; if
  // either is updated, both must be kept in sync.
  const impeller::ISize base_size = texture_->GetSize();
  const impeller::IRect destination_region = impeller::IRect::MakeXYWH(
      0, 0, MipDimensionAtLevel(base_size.width, mip_level),
      MipDimensionAtLevel(base_size.height, mip_level));

  // For the GLES backend, command queue submission just flushes the reactor,
  // which needs to happen on the raster thread.
  if (impeller_context.GetBackendType() ==
      impeller::Context::BackendType::kOpenGLES) {
    auto dart_state = flutter::UIDartState::Current();
    auto& task_runners = dart_state->GetTaskRunners();
    auto context_shared = gpu_context.GetContextShared();
    task_runners.GetRasterTaskRunner()->PostTask(fml::MakeCopyable(
        [context_shared, texture = texture_, staging_buffer, length,
         destination_region, mip_level, slice]() mutable {
          if (!EncodeAndSubmitOverwrite(*context_shared, texture,
                                        staging_buffer, length,
                                        destination_region, mip_level, slice)) {
            FML_LOG(ERROR) << "Failed to encode texture overwrite blit on the "
                              "raster thread.";
          }
          context_shared->DisposeThreadLocalCachedResources();
        }));
    return true;
  }

  if (!EncodeAndSubmitOverwrite(impeller_context, texture_, staging_buffer,
                                length, destination_region, mip_level, slice)) {
    return false;
  }
  impeller_context.DisposeThreadLocalCachedResources();
  return true;
}

#if IMPELLER_SUPPORTS_RENDERING
// Extracts the impeller::Texture that backs the ui.Image `image_wrapper`.
// Returns nullptr if the image is not backed by a Flutter GPU compatible
// texture, for example a non-Impeller image or a deferred image (from
// Picture.toImageSync) that has not finished rasterizing yet.
static std::shared_ptr<impeller::Texture> GetTextureFromImage(
    Context* gpu_context,
    Dart_Handle image_wrapper) {
  if (gpu_context == nullptr || Dart_IsNull(image_wrapper)) {
    return nullptr;
  }
  // A `ui.Image` wraps the private `_Image` native field holder.
  Dart_Handle inner = Dart_GetField(image_wrapper, tonic::ToDart("_image"));
  if (Dart_IsError(inner) || Dart_IsNull(inner)) {
    return nullptr;
  }
  auto* canvas_image =
      tonic::DartConverter<flutter::CanvasImage*>::FromDart(inner);
  if (!canvas_image) {
    return nullptr;
  }
  sk_sp<flutter::DlImage> dl_image = canvas_image->image();
  if (!dl_image) {
    return nullptr;
  }
  const impeller::DlImageImpeller* impeller_image = dl_image->asImpellerImage();
  if (!impeller_image) {
    return nullptr;
  }
  return impeller_image->GetImpellerTexture(gpu_context->GetContextShared());
}
#endif

Dart_Handle Texture::AsImage() const {
  // DlImageImpeller isn't compiled in builds with Impeller disabled. If
  // Impeller is disabled, it's impossible to get here anyhow, so just ifdef it
  // out.
#if IMPELLER_SUPPORTS_RENDERING
  auto image = flutter::CanvasImage::Create();
  auto dl_image = impeller::DlImageImpeller::Make(texture_);
  image->set_image(dl_image);
  auto wrapped = image->CreateOuterWrapping();
  return wrapped;
#else
  return Dart_Null();
#endif
}

}  // namespace gpu
}  // namespace flutter

//----------------------------------------------------------------------------
/// Exports
///

bool InternalFlutterGpu_Texture_Initialize(Dart_Handle wrapper,
                                           flutter::gpu::Context* gpu_context,
                                           int storage_mode,
                                           int format,
                                           int width,
                                           int height,
                                           int sample_count,
                                           int texture_type,
                                           bool enable_render_target_usage,
                                           bool enable_shader_read_usage,
                                           bool enable_shader_write_usage,
                                           int mip_level_count) {
  if (mip_level_count < 1) {
    return false;
  }
  impeller::TextureDescriptor desc;
  desc.storage_mode = flutter::gpu::ToImpellerStorageMode(storage_mode);
  desc.size = {width, height};
  desc.format = flutter::gpu::ToImpellerPixelFormat(format);
  desc.mip_count = static_cast<size_t>(mip_level_count);
  desc.usage = {};
  if (enable_render_target_usage) {
    desc.usage |= impeller::TextureUsage::kRenderTarget;
  }
  if (enable_shader_read_usage) {
    desc.usage |= impeller::TextureUsage::kShaderRead;
  }
  if (enable_shader_write_usage) {
    desc.usage |= impeller::TextureUsage::kShaderWrite;
  }
  switch (sample_count) {
    case 1:
      desc.sample_count = impeller::SampleCount::kCount1;
      break;
    case 4:
      desc.sample_count = impeller::SampleCount::kCount4;
      break;
    default:
      return false;
  }
  desc.type = static_cast<impeller::TextureType>(texture_type);
  if (!impeller::IsMultisampleCapable(desc.type) &&
      desc.sample_count != impeller::SampleCount::kCount1) {
    return false;
  }

  auto texture =
      gpu_context->GetContext().GetResourceAllocator()->CreateTexture(desc,
                                                                      true);
  if (!texture) {
    FML_LOG(ERROR) << "Failed to create texture.";
    return false;
  }

  auto res = fml::MakeRefCounted<flutter::gpu::Texture>(std::move(texture));
  res->AssociateWithDartWrapper(wrapper);

  return true;
}

bool InternalFlutterGpu_Texture_Overwrite(flutter::gpu::Texture* texture,
                                          flutter::gpu::Context* gpu_context,
                                          Dart_Handle source_byte_data,
                                          int mip_level,
                                          int slice) {
  if (mip_level < 0 || slice < 0) {
    return false;
  }
  return texture->Overwrite(*gpu_context, tonic::DartByteData(source_byte_data),
                            static_cast<uint32_t>(mip_level),
                            static_cast<uint32_t>(slice));
}

Dart_Handle InternalFlutterGpu_Texture_AsImage(flutter::gpu::Texture* wrapper) {
  return wrapper->AsImage();
}

Dart_Handle InternalFlutterGpu_Texture_ImageTextureInfo(
    flutter::gpu::Context* gpu_context,
    Dart_Handle image_wrapper) {
#if IMPELLER_SUPPORTS_RENDERING
  auto texture = flutter::gpu::GetTextureFromImage(gpu_context, image_wrapper);
  if (!texture) {
    return Dart_NewTypedData(Dart_TypedData_kInt32, 0);
  }
  const impeller::TextureDescriptor& desc = texture->GetTextureDescriptor();
  const impeller::TextureUsageMask usage = desc.usage;

  // Layout must match the parsing in the Dart `Texture._fromImage`.
  int32_t values[10];
  values[0] = static_cast<int32_t>(
      flutter::gpu::FromImpellerStorageMode(desc.storage_mode));
  values[1] =
      static_cast<int32_t>(flutter::gpu::FromImpellerPixelFormat(desc.format));
  values[2] = static_cast<int32_t>(desc.size.width);
  values[3] = static_cast<int32_t>(desc.size.height);
  values[4] = static_cast<int32_t>(desc.sample_count);
  // The Flutter GPU `TextureType` enum mirrors `impeller::TextureType`.
  values[5] = static_cast<int32_t>(desc.type);
  values[6] = (usage & impeller::TextureUsage::kRenderTarget) ? 1 : 0;
  values[7] = (usage & impeller::TextureUsage::kShaderRead) ? 1 : 0;
  values[8] = (usage & impeller::TextureUsage::kShaderWrite) ? 1 : 0;
  values[9] = static_cast<int32_t>(desc.mip_count);

  const intptr_t length = sizeof(values) / sizeof(values[0]);
  Dart_Handle list = Dart_NewTypedData(Dart_TypedData_kInt32, length);
  if (Dart_IsError(list)) {
    return list;
  }
  Dart_TypedData_Type type;
  void* data = nullptr;
  intptr_t data_length = 0;
  Dart_Handle acquire_result =
      Dart_TypedDataAcquireData(list, &type, &data, &data_length);
  if (Dart_IsError(acquire_result)) {
    return acquire_result;
  }
  std::memcpy(data, values, sizeof(values));
  Dart_TypedDataReleaseData(list);
  return list;
#else
  return Dart_NewTypedData(Dart_TypedData_kInt32, 0);
#endif
}

bool InternalFlutterGpu_Texture_InitializeFromImage(
    Dart_Handle wrapper,
    flutter::gpu::Context* gpu_context,
    Dart_Handle image_wrapper) {
#if IMPELLER_SUPPORTS_RENDERING
  auto texture = flutter::gpu::GetTextureFromImage(gpu_context, image_wrapper);
  if (!texture) {
    return false;
  }
  auto res = fml::MakeRefCounted<flutter::gpu::Texture>(std::move(texture));
  res->AssociateWithDartWrapper(wrapper);
  return true;
#else
  return false;
#endif
}
